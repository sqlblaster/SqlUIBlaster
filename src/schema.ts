export type Variable = {
  readonly id: Scalars["Primary"];
  readonly name: Scalars["String"];
  readonly order: Scalars["Int"];
  readonly type: VariableType;
  readonly model?: Maybe<Scalars["String"]>;
  readonly field?: Maybe<Scalars["String"]>;
  readonly value?: Maybe<Scalars["String"]>;
  readonly defaultValue?: Maybe<Scalars["String"]>;
//   readonly record?: Maybe<NavigationSearchResultItem>;
//   readonly defaultRecord?: Maybe<NavigationSearchResultItem>;
  readonly enums?: Maybe<ReadonlyArray<Scalars["String"]>>;
//   readonly widgets: ReadonlyArray<Widget>;
};

export enum VariableType {
  String = "STRING",
  Enum = "ENUM",
  Integer = "INTEGER",
  Float = "FLOAT",
  Date = "DATE",
  Record = "RECORD",
  Boolean = "BOOLEAN",
}

export type Maybe<T> = T | null | undefined;

export type Scalars = {
  ID: string;
  String: string;
  Boolean: boolean;
  Int: number;
  Float: number;
  /** Primary type */
  Primary: string;
  /** Timestamp type */
  Timestamp: Date;
  /** Foreign type */
  Foreign: string;
  /** Natural number type */
  Nat: number;
  /** The `JSON` scalar type represents JSON values as specified by [ECMA-404](http://www.ecma-international.org/publications/files/ECMA-ST/ECMA-404.pdf). */
  JSON: { [key: string]: any };
  /** Date type */
  Date: Date;
  /** Colour type */
  Colour: string;
  /** Uuid type */
  Uuid: string;
  DateTime: Date;
};

export type GraphNode = {
  readonly id: Scalars["Primary"];
  readonly rawId: Scalars["String"];
  readonly title?: Maybe<Scalars["String"]>;
  readonly type: Scalars["String"];
  readonly through: Scalars["Boolean"];
  readonly glyph?: Maybe<Scalars["String"]>;
};

export type NavigationSearchResultItem = {
  readonly id: Scalars["String"];
  readonly url: Scalars["String"];
  readonly type: Scalars["String"];
  readonly title?: Maybe<Scalars["String"]>;
  readonly model: Scalars["String"];
  readonly headline: Scalars["String"];
  readonly rank: Scalars["Float"];
  readonly glyph?: Maybe<Scalars["String"]>;
};
