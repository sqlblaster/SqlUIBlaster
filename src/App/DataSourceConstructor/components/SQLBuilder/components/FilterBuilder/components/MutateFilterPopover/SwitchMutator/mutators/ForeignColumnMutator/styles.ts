import { StyleRulesCallback } from '@material-ui/core';

export type ForeignColumnMutatorClassKeys =
  | 'root'
  | 'common-padding'
  | 'title-and-operator'
  | 'title'
  | 'mutator-action'
  | 'hr';

export const foreignColumnMutatorStyles: StyleRulesCallback<
  ForeignColumnMutatorClassKeys
> = (theme) => ({
  root: {
    border: '1px solid #EBECED',
    borderRadius: '5px'
  },
  'common-padding': {
    padding: 10
  },
  'title-and-operator': {
    display: 'flex',
    alignItems: 'center',
    justifyContent: 'space-between'
  },
  title: {
    marginRight: 20
  },
  'mutator-action': {
    display: 'flex',
    alignItems: 'center',
    justifyContent: 'flex-end',
    padding: '5px'
  },
  hr: {
    margin: '0px',
    background: '#EBECED',
    height: '1px',
    border: 'none'
  }
});
