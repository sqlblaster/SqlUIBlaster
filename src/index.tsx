import React from 'react';
import ReactDOM from 'react-dom';
import { DataSourceConstructor } from './App/DataSourceConstructor';

ReactDOM.render(
  <React.StrictMode>
    <DataSourceConstructor databaseSchema={{ tables:[] }}/>
  </React.StrictMode>,
  document.getElementById('root')
);
