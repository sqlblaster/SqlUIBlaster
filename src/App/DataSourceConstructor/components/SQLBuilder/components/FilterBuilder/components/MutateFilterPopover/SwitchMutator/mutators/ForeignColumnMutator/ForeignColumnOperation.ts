import { GraphNode } from 'src/schema';
import { IOperation, Operation } from '../../../Operation';

export class ForeignColumnOperation implements IOperation<'Foreign', ForeignColumnOperand> {
  public operator: 'ForeignColumnOperator' = 'ForeignColumnOperator';
  public operands: ForeignColumnOperand;
  public type: 'Foreign' = 'Foreign';

  constructor(...inits: Partial<ForeignColumnOperation>[]) {
    Object.assign(this, ...inits);
  }
}

export type ForeignColumnOperand = GraphNode | undefined;
