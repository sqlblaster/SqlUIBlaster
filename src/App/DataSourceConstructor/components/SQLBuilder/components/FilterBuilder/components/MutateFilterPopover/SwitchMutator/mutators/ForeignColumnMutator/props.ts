import { ClassesProp } from 'src/components/query-builder/App/utils/classes-prop';
import {
  actionCreators,
  PickActionCreators
} from 'src/components/query-builder/store/action-creators';
import { PickStates } from 'src/components/query-builder/store/models/State';
import { ForeignFieldProps } from 'src/process/startnew/ForeignField';
import { VariablesPickerProps } from '../../VariablesPicker/props';
import { MutatorProps } from '../props';
import { ForeignColumnOperation } from './ForeignColumnOperation';
import { ForeignColumnMutatorClassKeys } from './styles';

export type ForeignColumnMutatorOwnProps = MutatorProps;

export type ForeignColumnMutatorStateProps = {
  record: ForeignColumnOperation['operands'];
  enums: string[] | undefined;
} & PickStates<'canSaveFilter' | 'pickedVariables'>;

export type ForeignColumnMutatorDispatchProps = PickActionCreators<
  'setOperation' | 'changeCanSaveFilterState' | 'setPickedVariables'
>;

const { setOperation, setPickedVariables, changeCanSaveFilterState } = actionCreators;
export const FCMMapDispatchToProps: ForeignColumnMutatorDispatchProps = {
  setOperation,
  setPickedVariables,
  changeCanSaveFilterState
};

export type ForeignColumnMutatorProps = ForeignColumnMutatorOwnProps &
  ForeignColumnMutatorStateProps &
  ForeignColumnMutatorDispatchProps;

export interface IForeignColumnMutatorHandlers {
  handleOperandChange: ForeignFieldProps['handleChange'];
  handleVariablePick: VariablesPickerProps['onVariablePicked'];
  filter: Required<VariablesPickerProps>['filter'];
}

export type ForeignColumnMutatorViewProps = ForeignColumnMutatorOwnProps &
  ForeignColumnMutatorStateProps &
  IForeignColumnMutatorHandlers &
  ClassesProp<ForeignColumnMutatorClassKeys>;
