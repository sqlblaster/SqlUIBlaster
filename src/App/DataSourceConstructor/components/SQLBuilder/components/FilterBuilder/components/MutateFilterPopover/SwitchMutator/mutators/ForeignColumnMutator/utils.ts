import { ForeignColumnMutatorProps } from './props';

export const getModels = (
  props: Pick<ForeignColumnMutatorProps, 'enums' | 'column'>
): string[] | undefined => {
  const {
    enums,
    column: {
      lastColumn: { foreignModelName, modelName, type }
    }
  } = props;

  return type === 'uuid' && !enums
    ? undefined
    : type === 'uuid' && enums
    ? enums
    : type === 'primary'
    ? [modelName]
    : foreignModelName
    ? [foreignModelName]
    : undefined;
};
