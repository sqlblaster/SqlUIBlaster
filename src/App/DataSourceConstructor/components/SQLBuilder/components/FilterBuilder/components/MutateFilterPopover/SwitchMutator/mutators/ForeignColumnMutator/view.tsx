import { withStyles } from '@material-ui/core';
import * as React from 'react';
import { ForeignField } from 'src/process/startnew/ForeignField';
import { VariableType } from 'src/schema';
import { VariableItem } from '../../VariableItem';
import { VariablesPicker } from '../../VariablesPicker';
import { ForeignColumnMutatorViewProps } from './props';
import { foreignColumnMutatorStyles } from './styles';
import { getModels } from './utils';

export const ForeignColumnMutator = withStyles(foreignColumnMutatorStyles)(
  class extends React.Component<ForeignColumnMutatorViewProps> {
    public render() {
      const {
        record,
        pickedVariables,
        titleComponent,
        classes,
        submitMutationComponent,
        filter,
        handleOperandChange,
        handleVariablePick
      } = this.props;

      return (
        <div className={classes.root}>
          <div
            className={`${classes['title-and-operator']} ${classes['common-padding']}`}
          >
            <div className={classes.title}>{titleComponent}</div>
          </div>
          <hr className={classes.hr} />
          <div className={classes['common-padding']}>
            <ForeignField
              handleChange={handleOperandChange}
              id={record ? record.id : ''}
              parent={
                pickedVariables && !(pickedVariables instanceof Array)
                  ? { title: '', type: pickedVariables.model as string }
                  : record
              }
              label={''}
              models={getModels(this.props)}
              endAdornment={
                <VariablesPicker
                  type={VariableType.Record}
                  filter={filter}
                  onVariablePicked={handleVariablePick}
                />
              }
              startAdornment={
                pickedVariables &&
                !(pickedVariables instanceof Array) && (
                  <VariableItem variable={pickedVariables} />
                )
              }
            />
          </div>
          <hr className={classes.hr} />
          <div
            className={`${classes['mutator-action']} ${classes['common-padding']}`}
          >
            {submitMutationComponent}
          </div>
        </div>
      );
    }
  }
);
