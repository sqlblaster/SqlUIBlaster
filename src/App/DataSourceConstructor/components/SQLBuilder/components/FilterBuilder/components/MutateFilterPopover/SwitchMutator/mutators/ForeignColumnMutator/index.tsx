import * as React from 'react';
import { connect } from 'react-redux';
import { State } from 'src/components/query-builder/store/models/State';
import { GraphNode } from 'src/schema';
import { getEnums } from '../utils';
import { ForeignColumnOperation } from './ForeignColumnOperation';
import {
  FCMMapDispatchToProps,
  ForeignColumnMutatorDispatchProps,
  ForeignColumnMutatorOwnProps,
  ForeignColumnMutatorProps,
  ForeignColumnMutatorStateProps,
  ForeignColumnMutatorViewProps as ViewProps,
  IForeignColumnMutatorHandlers
} from './props';
import { getModels } from './utils';
import { ForeignColumnMutator as View } from './view';

export const ForeignColumnMutator = connect<
  ForeignColumnMutatorStateProps,
  ForeignColumnMutatorDispatchProps,
  ForeignColumnMutatorOwnProps,
  State
>(
  ({ operation, canSaveFilter, pickedVariables, tables }, { column }) => {
    const foreignColumnOpertion =
      operation.type === 'Foreign'
        ? new ForeignColumnOperation(operation)
        : new ForeignColumnOperation();

    return {
      record: foreignColumnOpertion.operands,
      canSaveFilter,
      pickedVariables,
      enums: getEnums(tables, column)
    };
  },
  FCMMapDispatchToProps
)(
  class extends React.Component<ForeignColumnMutatorProps>
    implements IForeignColumnMutatorHandlers {
    public componentDidMount() {
      const { record, setOperation, changeCanSaveFilterState } = this.props;
      setOperation(new ForeignColumnOperation({ operands: record }));
      changeCanSaveFilterState(!!record);
    }

    public componentWillUnmount() {
      this.props.changeCanSaveFilterState(true);
    }

    public handleOperandChange: ViewProps['handleOperandChange'] = record => {
      const {
        setOperation,
        setPickedVariables,
        changeCanSaveFilterState
      } = this.props;

      changeCanSaveFilterState(!!record);
      setOperation(
        new ForeignColumnOperation({ operands: record as GraphNode })
      );
      setPickedVariables(null);
    }

    public handleVariablePick: ViewProps['handleVariablePick'] = variable => {
      const { setPickedVariables, changeCanSaveFilterState } = this.props;
      setPickedVariables(variable);
      changeCanSaveFilterState(true);
    }

    public filter: ViewProps['filter'] = variable => {
      const models = getModels(this.props);

      return !models || models.some(model => model === variable.model);
    }

    public render() {
      return (
        <View
          handleOperandChange={this.handleOperandChange}
          handleVariablePick={this.handleVariablePick}
          filter={this.filter}
          {...this.props}
        />
      );
    }
  }
);
