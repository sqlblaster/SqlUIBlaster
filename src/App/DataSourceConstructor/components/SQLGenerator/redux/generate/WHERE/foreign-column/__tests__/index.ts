import { ForeignColumnOperation } from 'src/components/query-builder/App/DataSourceConstructor/components/SQLBuilder/components/FilterBuilder/components/MutateFilterPopover/SwitchMutator/mutators/ForeignColumnMutator/ForeignColumnOperation';
import { Variable, VariableType } from 'src/schema';
import { generateForeignColumnOperation } from '..';
import {
  ColumnBranchAliased,
  FilterAliased
} from '../../../JOIN/models/ColumnAliased';

describe('foreign column filter generator', () => {
  it('should render properly when there is picked record', () => {
    expect(
      generateForeignColumnOperation(
        new FilterAliased({
          column: new ColumnBranchAliased({
            columnName: 'account_id',
            tableName: 'balance',
            type: 'string',
            foreignTableName: 'Account',
            tableAlias: 'balance1'
          }),
          operation: new ForeignColumnOperation({
            operands: {
              id: '59da5627-cb22-4b4e-884c-ad8cf94ebd4e',
              type: 'account',
              glyph: null,
              title: 'Feeney',
              rawId: '',
              through: true
            }
          }),
          id: 'be2651d6-82bd-41ff-b272-bb4288cf3ce3'
        })
      )
    ).toMatchSnapshot();
  });

  it('should render properly when there is picked variable', () => {
    const variables: Variable = {
      id: '584EVVN5j8j6UrP6FXW3zm',
      name: 'Country',
      type: VariableType.Record,
      order: 2,
      model: 'Country',
      widgets: []
    };

    expect(
      generateForeignColumnOperation(
        new FilterAliased({
          column: new ColumnBranchAliased({
            columnName: 'country_id',
            tableName: 'employee',
            type: 'string',
            foreignTableName: 'Country',
            tableAlias: 'employee1'
          }),
          operation: new ForeignColumnOperation({
            operands: {
              id: 'a42791bc-9149-4880-8b26-61aa51278354',
              type: 'country',
              glyph: null,
              title: 'Kazakhstan',
              rawId: '',
              through: true
            }
          }),
          variables,
          id: 'd541c88f-a968-4254-a04d-79b67bfef530'
        })
      )
    ).toMatchSnapshot();
  });
});
