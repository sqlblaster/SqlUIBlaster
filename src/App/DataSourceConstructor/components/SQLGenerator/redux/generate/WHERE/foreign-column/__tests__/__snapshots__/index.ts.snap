// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`foreign column filter generator should render properly when there is picked record 1`] = `"\\"balance1\\".\\"account_id\\" = '59da5627-cb22-4b4e-884c-ad8cf94ebd4e'"`;

exports[`foreign column filter generator should render properly when there is picked variable 1`] = `"\\"employee1\\".\\"country_id\\" = $2::uuid"`;
